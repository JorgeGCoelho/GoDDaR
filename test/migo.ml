open Dlock
open Dlock.Types;;

let readdir dir: string list = 
  Sys.readdir dir 
  |> Array.map (fun subdir -> (dir ^ "/" ^ subdir))
  |> Array.to_list
;;

(* Format.printf "%s\n" (Sys.getcwd()); *)
Sys.chdir "data";
[(readdir "simple"); (readdir "benchmark")]
|>  List.flatten
|>  List.filter (fun elem -> (Sys.is_directory elem) && (String.get elem 0 <> '.'))
|>  (List.sort compare)
|>  List.iter (
      fun dir -> 
        let migo_file = (dir ^ "/main.migo" ) in
        Format.printf "%s\n" migo_file;
        if Sys.file_exists migo_file then (
          (
            try 
              let fmt = Format.std_formatter in
              let migo = MiGo.parse_file migo_file in
              MiGo_Types.print_migo_list fmt migo;
              Format.printf "%s\n" (String.make 10 '=');
              (
                try 
                  let (ccs, deps) = (MiGo_to_CCS.migo_to_ccs migo) in
                  Lambda.print_lambda_simple fmt (Types.lambdaTaggedToLambda ccs);
                  Format.printf "\n";
                  if deps <> [] then (
                    Format.fprintf fmt "DEPS:\n"; 
                    List.iter (fun dep -> Format.fprintf fmt "%a\n" Types.print_dependecy dep) deps;
                  )
                with
                | MiGo_to_CCS.Fail(reason) -> Format.printf "FAILED TO CONVERT: %s\n" reason;
              );
            with
            | _ -> Format.printf "FAILED TO PARSE MIGO\n";
          );
          Format.printf "%s\n" (String.make 50 '=');
        );
    )