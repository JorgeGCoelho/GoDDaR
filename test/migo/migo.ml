open Dlock;;

(* Format.printf "%s\n" (Sys.getcwd()); *)
Sys.chdir "data";
( Array.to_list (Sys.readdir "."))
|> List.filter (
  fun elem -> 
    (Sys.is_directory elem) && (String.get elem 0 <> '.')
) |> (
  List.sort compare
) |> List.iter (
  fun dir -> 
    let migo_file = (dir ^ "/main.migo" ) in
    Format.printf "%s\n" migo_file;
    let migo = MiGo.parse_file migo_file in
    MiGo_Types.print_migo_list Format.std_formatter migo;
    Format.printf "%s\n" (String.make 10 '=');
    (
      try 
        let ccs = (MiGo_to_CCS.migo_to_ccs migo) in
        Format.printf "%s\n" ccs;
        ignore (CCS.parse ccs);
      with
      | MiGo_to_CCS.Fail(reason) -> Format.printf "FAILED TO CONVERT: %s\n" reason;
      | _ -> Format.printf "NOT VALID CCS"
    );
    Format.printf "%s\n" (String.make 50 '=');
)