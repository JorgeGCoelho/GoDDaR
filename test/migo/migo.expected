if-send-or-receive/main.migo
def main.main():
    let t0 = newchan main.main0.t0_chan0, 0;
    if
        call main.main#1(t0);
    else
        call main.main#3(t0);
    endif;
def main.main#1(t0):
    send t0;
def main.main#3(t0):
    recv t0;
==========
(t0!.0 + t0?.0)
==================================================
loop-send/main.migo
def main.main():
    let t0 = newchan main.main0.t0_chan0, 0;
    call main.main#1(t0);
def main.main#1(t0):
    send t0;
    call main.main#1(t0);
==========
FAILED TO CONVERT: Recursive call
==================================================
recv/main.migo
def main.main():
    let t0 = newchan main.main0.t0_chan0, 0;
    recv t0;
==========
t0?.0
==================================================
send/main.migo
def main.main():
    let t0 = newchan main.main0.t0_chan0, 0;
    send t0;
==========
t0!.0
==================================================
send-recv/main.migo
def main.main():
    let t0 = newchan main.main0.t0_chan0, 0;
    recv t0;
    send t0;
==========
t0?.t0!.0
==================================================
two-if-send-recv-func/main.migo
def main.main():
    let t0 = newchan main.main0.t0_chan0, 0;
    call main.send(t0);
    call main.send(t0);
def main.send(a):
    if
        call main.send#1(a);
    else
        call main.send#3(a);
    endif;
def main.send#1(a):
    send a;
def main.send#3(a):
    recv a;
==========
(a!.(a!.0 + a?.0) + a?.(a!.0 + a?.0))
==================================================
two-if-send-recv-func2/main.migo
def main.main():
    let t0 = newchan main.main0.t0_chan0, 0;
    let t1 = newchan main.main0.t1_chan0, 0;
    call main.send(t0, t1);
    call main.send(t1, t0);
def main.send(a, b):
    if
        call main.send#1(a, b);
    else
        call main.send#3(a, b);
    endif;
def main.send#1(a, b):
    send a;
def main.send#3(a, b):
    recv b;
==========
(a!.(a!.0 + b?.0) + b?.(a!.0 + b?.0))
==================================================
two-send-func/main.migo
def main.main():
    let t0 = newchan main.main0.t0_chan0, 0;
    call main.send(t0);
    call main.send(t0);
def main.send(a):
    send a;
==========
a!.a!.0
==================================================
